### PR TITLE
92 add testing documentation to the project

### LINKED ISSUE #92: Add Testing Documentation to the Project
Description:

The purpose of this issue is to  prepare a comprehensive testing documentation file to guide contributors in writing, organising, and running tests consistently. The document should include testing principles, test types, writing guidelines, and CI integration practices.

Tasks:
- [ ] Create a new file TESTING.md in the project root.

 

- [ ] Add sections for:

- Testing Principles
- Test Types (Unit, Integration, Edge Case, Regression)
- Writing Tests (Structure and Checklist)
- Running Tests
- Tools and Frameworks
- Continuous Integration

 

- [ ] Verify all instructions are accurate and applicable to the project.

 

- [ ] Add example configurations for CI and test coverage.

 

- [ ] Submit the document for review.

References:

- Refer to the draft document prepared in the `TESTING.md`

### PR DESCRIPTION
**Pull request by tnzmnjm**
Add testing documentation and update PR template (#92)

**Scope:** 
Establish comprehensive testing guidelines and integrate testing requirements into the PR workflow.

**Description of changes:**
- Created `TESTING.md` with complete testing documentation including:
  - Testing principles and best practices
  - Test types (unit, integration, edge case, regression)
  - Test writing guidelines and structure
  - Instructions for running tests and viewing coverage
  - CI/CD configuration examples with GitHub Actions
  - Coverage threshold recommendations

- Updated `pull_request_template.md` to:
  - Add dedicated Testing section requiring test documentation
  - Include reference to TESTING.md for guidance
  - Extend checklist with test-specific requirements
  - Ensure contributors confirm tests for new functionality
  - Verify existing tests continue to pass

**Checklist:**
- [x] Follows coding standards
- [x] Self-reviewed
- [x] Comments explain difficult code
- [x] Updated documentation
- [x] No new warnings
- [x] Dependent changes merged
- [x] Tests added for new functionality
- [x] Existing tests pass without modifications

**Testing:**
- **Tests added:** N/A (Documentation changes only)
- **Test coverage:** N/A
- **Test cases reviewed:** N/A

**References:**
- [x] Issue #92 - Add testing documentation to the project

### CODE REVIEW

File: TESTING.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+```bash
+pytest
+```
+
+### Running Specific Tests
+
+- To run a specific test file:
+  ```bash
+  pytest tests/test_file_name.py
```
Comment (julianharty) [inline TESTING.md:103]: These are general instructions. Could you tailor them to call one of the project's test files so that someone can copy and paste this example and actually run the tests in one of the project's files please?

File: TESTING.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  ```
+- To run a specific test case:
+  ```bash
+  pytest tests/test_file_name.py::test_case_name
```
Comment (julianharty) [inline TESTING.md:107]: As per my previous comment, it'd be good to tailor the example to run a specific test case that exists in this project.

File: TESTING.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  ```
+
+### Viewing Test Coverage
+
+Generate a coverage report using `pytest-cov`:
+
+```bash
+pytest --cov=your_project tests/
```
Comment (julianharty) [inline TESTING.md:115]: Ditto - could you change this to use a suitable project name for this repo?

**Review by julianharty — CHANGES_REQUESTED** [review]
Looks good overall, however please tailor the docs to this project.

File: TESTING.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
+# Testing Documentation
+
+This document provides guidelines for writing, organising, and running tests
```
[the remark this answers is not in the document]
Comment (julianharty) [reply TESTING.md:3]: I'd prefer to keep British English spellings.

File: TESTING.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+- **pytest:** Primary testing framework for this project.
+- **unittest:** Alternative framework for writing basic test cases.
+
+### Mocking Tools
+
+- **unittest.mock:** Mock dependencies and isolate units for testing.
+
+### Coverage Tools
+
+- **pytest-cov:** Analyse code coverage to ensure all parts of the code are tested.
```
[the remark this answers is not in the document]
Comment (julianharty) [reply TESTING.md:133]: I'd prefer to keep British English spellings.

File: TESTING.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+- **Examples:**
+  - Handling empty, null, or unexpected input types.
+  - Large-scale data processing.
+
+### Regression Tests
+
+- **Definition:** Test previously fixed bugs to ensure they do not reappear.
+- **Purpose:** Maintain stability as new changes are introduced.
+- **Examples:**
+  - Re-testing a bug fix in data serialisation logic.
```
[the remark this answers is not in the document]
Comment (julianharty) [reply TESTING.md:71]: I'd prefer to keep British English spellings.
